dhall-toml/src/Dhall/Toml.hs

```diff
@@ -25,68 +29,134 @@ import qualified Dhall.Parser
+    -- is alloed by the spec
+    | UnsupportedArray (Expr Void Void)
     | NotARecord (Expr Void Void)
+    -- the latest TOML spec, v1.0.0 allows this but tomland has not
+    -- implemented it yet
+    -- NOTE: the only way to get this error is through enums
+    | HeterogeneousArray (Expr Void Void)
 
 instance Show CompileError where
     show (Unimplemented s) = "unimplemented: " ++ s
     show (Unsupported e) = "unsupported: " ++ show e
+    show (UnsupportedArray e) = "records cannot be nested in multi-dimentional arrays: " ++ show e
     show (NotARecord e) = "The root object converted to TOML must be a record, got " ++ show e
+    show (HeterogeneousArray e) = "heterogeneous arrays are not currently supported " ++ show e
 
 instance Exception CompileError
 
 dhallToToml :: Expr s Void -> Either CompileError TOML
 dhallToToml e0 = do
-    let norm = Core.normalize e0
-    _ <- assertRecordLit norm
-    toToml (mempty :: TOML) [] norm
+    r <- assertRecordLit (Core.normalize e0)
+    toTomlTable r
+
+assertRecordLit :: Expr Void Void -> Either CompileError (Map.Map Text (Core.RecordField Void Void))
+assertRecordLit (Core.RecordLit r) = Right r
+assertRecordLit e                  = Left $ NotARecord e
+
+toTomlTable :: Map.Map Text (Core.RecordField Void Void) -> Either CompileError TOML
+toTomlTable r = foldM (toTomlRecordFold []) (mempty :: TOML) (Map.toList r)
+
+toTomlRecordFold :: [Piece] -> TOML -> (Text, Core.RecordField Void Void) -> Either CompileError TOML
+toTomlRecordFold curKey toml' (key', val) = toToml toml' newKey (Core.recordFieldValue val)
     where
-        assertRecordLit (Core.RecordLit r) = Right r
-        assertRecordLit e                  = Left $ NotARecord e
-
--- | A helper function for dhallToToml. It recursively adds the values in
---   the Expr to the TOML. It has an invariant that key can be null iff
---   Expr is a RecordLit. This aligns with how a TOML document must be a table,
---   and bare values cannot be represented
-toToml :: TOML -> [Piece] -> Expr Void Void -> Either CompileError TOML
+        append []     y = y :| []
+        append (x:xs) y = x :| xs ++ [y]
+        newKey = Key $ append curKey $ Piece key'
+
+toToml :: TOML -> Key -> Expr Void Void -> Either CompileError TOML
 toToml toml key expr  = case expr of
     Core.BoolLit a -> return $ insertPrim (Toml.Value.Bool a)
     Core.NaturalLit a -> return $ insertPrim (Toml.Value.Integer $ toInteger a)
     Core.DoubleLit (DhallDouble a) -> return $ insertPrim (Toml.Value.Double a)
     Core.TextLit (Core.Chunks [] a) -> return $ insertPrim (Toml.Value.Text a)
-    -- TODO: probe the element type, if record then table list else inline list
-    -- Core.ListLit _ a -> Left $ Unimplemented
+    Core.ListLit _ a -> case toList a of-- Seq.lookup 0 a of
+        -- empty array
+        [] -> return $ insertPrim (Toml.Value.Array [])
+        -- array of table
+        record@(Core.RecordLit _) : records -> do
```

Comment:
As an alternative to converting to lists, you could use [`Seq`'s bundled pattern synonyms `Empty` and `:<|`](https://hackage.haskell.org/package/containers-0.6.5.1/docs/Data-Sequence.html#t:Seq) here.

Reply:
That's good to know! But, this ends up getting converted to a list for the latter to cases (`insertTableArrays` needs a `NonEmpty` and `toMArray` needs a plain `[]`) so I think it's cleaner with `toList` at the top of the `case`.  This made me realize the the last case can be
```haskell
a' -> do
    anyList <- mapM toAny a'
```

instead of calling `toList` again.